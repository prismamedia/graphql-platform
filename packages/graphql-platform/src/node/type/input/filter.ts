import * as scalars from '@prismamedia/graphql-platform-scalars';
import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/memoize';
import type { Edge, Node } from '../../../node.js';
import { Leaf, type LeafValue } from '../../definition/component/leaf.js';
import type { MultipleReverseEdge } from '../../definition/reverse-edge/multiple.js';
import { UniqueReverseEdge } from '../../definition/reverse-edge/unique.js';
import type { OperationContext } from '../../operation/context.js';
import {
  AndOperation,
  BooleanFilter,
  EdgeExistsFilter,
  FalseValue,
  LeafComparisonFilter,
  LeafFullTextFilter,
  LeafInFilter,
  MultipleReverseEdgeCountFilter,
  MultipleReverseEdgeExistsFilter,
  NodeFilter,
  NotOperation,
  OrOperation,
  TrueValue,
  UniqueReverseEdgeExistsFilter,
  sortableLeafComparisonOperatorSet,
} from '../../statement/filter.js';
import {
  BooleanOperationFilterInput,
  EdgeFilterInput,
  LeafFilterInput,
  ReverseEdgeFilterInput,
  type FieldFilterInputInterface,
} from './filter/field.js';

export * from './filter/field.js';

export type NodeFilterInputValue = utils.Nillable<utils.PlainObject>;

export type NodeFilterInputTypeOverride = {
  name?: string;
  description?: string;
};

export class NodeFilterInputType extends utils.ObjectInputType<FieldFilterInputInterface> {
  public static createLeafComparisonFields(leaf: Leaf): LeafFilterInput[] {
    const fields: LeafFilterInput[] = [];

    const inputType = scalars.stringTypes.includes(leaf.type as any)
      ? scalars.typesByName.NonEmptyTrimmedString
      : leaf.type;

    if (leaf.isComparable()) {
      // eq, not
      for (const operator of ['eq', 'not'] as const) {
        fields.push(
          new LeafFilterInput<LeafValue>(leaf, operator, {
            type: utils.nonNullableInputTypeDecorator(
              inputType,
              !leaf.isNullable(),
            ),
            filter: (value, _context, _path) =>
              new LeafComparisonFilter(leaf, operator, value),
          }),
        );
      }

      if (leaf.type !== scalars.typesByName.Boolean || leaf.isNullable()) {
        // in, not_in
        fields.push(
          new LeafFilterInput<LeafValue[]>(leaf, 'in', {
            type: new utils.NonNullableInputType(
              new utils.ListableInputType(
                new utils.NonOptionalInputType(
                  utils.nonNullableInputTypeDecorator(
                    inputType,
                    !leaf.isNullable(),
                  ),
                ),
              ),
            ),
            filter: (values, _context, _path) =>
              LeafInFilter.create(leaf, values),
          }),
          new LeafFilterInput<LeafValue[]>(leaf, 'not_in', {
            type: new utils.NonNullableInputType(
              new utils.ListableInputType(
                new utils.NonOptionalInputType(
                  utils.nonNullableInputTypeDecorator(
                    inputType,
                    !leaf.isNullable(),
                  ),
                ),
              ),
            ),
            filter: (values, _context, _path) =>
              NotOperation.create(LeafInFilter.create(leaf, values)),
          }),
        );
      }

      // gt, gte, lt, lte
      if (leaf.isSortable()) {
        for (const operator of sortableLeafComparisonOperatorSet) {
          fields.push(
            new LeafFilterInput<NonNullable<LeafValue>>(leaf, operator, {
              type: new utils.NonNullableInputType(inputType),
              filter: (value, _context, _path) =>
                new LeafComparisonFilter(leaf, operator, value),
            }),
          );
        }
      }
    }

    // is_null
    if (leaf.isNullable()) {
      fields.push(
        new LeafFilterInput<boolean>(leaf, 'is_null', {
          type: new utils.NonNullableInputType(scalars.typesByName.Boolean),
          filter: (value, _context, _path) =>
            new LeafComparisonFilter(leaf, value ? 'eq' : 'not', null),
        }),
      );
    }

    return fields;
  }

  public static createLeafFullTextFields(leaf: Leaf): LeafFilterInput[] {
    const fields: LeafFilterInput[] = [];

    // Full-text search
    // contains, starts_with, ends_with
    if (
      [
        scalars.typesByName.DraftJS,
        scalars.typesByName.URL,
        ...scalars.stringTypes,
      ].includes(leaf.type as any)
    ) {
      for (const operator of [
        'contains',
        'starts_with',
        'ends_with',
      ] as const) {
        fields.push(
          new LeafFilterInput<string>(leaf, operator, {
            type: new utils.NonNullableInputType(
              scalars.typesByName.NonEmptyString,
            ),
            filter: (value, _context, _path) =>
              new LeafFullTextFilter(leaf, operator, value),
          }),
          new LeafFilterInput<string>(leaf, `not_${operator}`, {
            type: new utils.NonNullableInputType(
              scalars.typesByName.NonEmptyString,
            ),
            filter: (value, _context, _path) =>
              NotOperation.create(
                new LeafFullTextFilter(leaf, operator, value),
              ),
          }),
        );
      }
    }

    return fields;
  }

  public static createLeafFields(leaf: Leaf): LeafFilterInput[] {
    return [
      ...this.createLeafComparisonFields(leaf),
      ...this.createLeafFullTextFields(leaf),
    ];
  }

  public static createEdgeFields(
    edge: Edge,
    headFilterInputType: NodeFilterInputType = edge.head.filterInputType,
  ): EdgeFilterInput[] {
    const fields: EdgeFilterInput[] = [
      new EdgeFilterInput<NodeFilterInputValue>(edge, 'eq', {
        type: utils.nonNullableInputTypeDecorator(
          headFilterInputType,
          !edge.isNullable(),
        ),
        filter: (value, context, path) =>
          value === null
            ? NotOperation.create(EdgeExistsFilter.create(edge))
            : EdgeExistsFilter.create(
                edge,
                headFilterInputType.filter(value, context, path),
              ),
      }),
      new EdgeFilterInput<NodeFilterInputValue>(edge, 'not', {
        type: utils.nonNullableInputTypeDecorator(
          headFilterInputType,
          !edge.isNullable(),
        ),
        filter: (value, context, path) =>
          value === null
            ? EdgeExistsFilter.create(edge)
            : NotOperation.create(
                EdgeExistsFilter.create(
                  edge,
                  headFilterInputType.filter(value, context, path),
                ),
              ),
      }),
    ];

    // is_null
    if (edge.isNullable()) {
      fields.push(
        new EdgeFilterInput<boolean>(edge, 'is_null', {
          type: new utils.NonNullableInputType(scalars.typesByName.Boolean),
          filter: (value, _context, _path) =>
            value
              ? NotOperation.create(EdgeExistsFilter.create(edge))
              : EdgeExistsFilter.create(edge),
        }),
      );
    }

    return fields;
  }

  public static createUniqueReverseEdgeFields(
    reverseEdge: UniqueReverseEdge,
    headFilterInputType: NodeFilterInputType = reverseEdge.head.filterInputType,
  ): ReverseEdgeFilterInput[] {
    return [
      new ReverseEdgeFilterInput<NodeFilterInputValue>(reverseEdge, 'eq', {
        type: headFilterInputType,
        filter: (value, context, path) =>
          value === null
            ? NotOperation.create(
                UniqueReverseEdgeExistsFilter.create(reverseEdge),
              )
            : UniqueReverseEdgeExistsFilter.create(
                reverseEdge,
                headFilterInputType.filter(value, context, path),
              ),
      }),
      new ReverseEdgeFilterInput<NodeFilterInputValue>(reverseEdge, 'not', {
        type: headFilterInputType,
        filter: (value, context, path) =>
          value === null
            ? UniqueReverseEdgeExistsFilter.create(reverseEdge)
            : NotOperation.create(
                UniqueReverseEdgeExistsFilter.create(
                  reverseEdge,
                  headFilterInputType.filter(value, context, path),
                ),
              ),
      }),
      new ReverseEdgeFilterInput<NodeFilterInputValue>(reverseEdge, 'is_null', {
        type: new utils.NonNullableInputType(scalars.typesByName.Boolean),
        filter: (value, _context, _path) =>
          value
            ? NotOperation.create(
                UniqueReverseEdgeExistsFilter.create(reverseEdge),
              )
            : UniqueReverseEdgeExistsFilter.create(reverseEdge),
      }),
    ];
  }

  /**
   * We can note the following equivalences:
   *
   *  set.every(filter);
   *    = !set.some(!filter);
   *    = set.none(!filter);
   *
   *  set.some(filter);
   *    = !set.every(!filter);
   *    = !set.none(filter);
   *
   *  set.none(filter);
   *    = !set.some(filter);
   *    = set.every(!filter);
   */
  public static createMultipleReverseEdgeFields(
    reverseEdge: MultipleReverseEdge,
    headFilterInputType: NodeFilterInputType = reverseEdge.head.filterInputType,
  ): ReverseEdgeFilterInput[] {
    return [
      new ReverseEdgeFilterInput<NonNullable<NodeFilterInputValue>>(
        reverseEdge,
        'every',
        {
          type: new utils.NonNullableInputType(headFilterInputType),

          // set.every(filter) = !set.some(!filter);
          filter: (value, context, path) =>
            NotOperation.create(
              MultipleReverseEdgeExistsFilter.create(
                reverseEdge,
                headFilterInputType.filter(value, context, path).complement,
              ),
            ),
        },
      ),
      new ReverseEdgeFilterInput<NonNullable<NodeFilterInputValue>>(
        reverseEdge,
        'some',
        {
          type: new utils.NonNullableInputType(headFilterInputType),
          filter: (value, context, path) =>
            MultipleReverseEdgeExistsFilter.create(
              reverseEdge,
              headFilterInputType.filter(value, context, path),
            ),
        },
      ),
      new ReverseEdgeFilterInput<NonNullable<NodeFilterInputValue>>(
        reverseEdge,
        'none',
        {
          type: new utils.NonNullableInputType(headFilterInputType),

          // set.none(filter) = !set.some(filter);
          filter: (value, context, path) =>
            NotOperation.create(
              MultipleReverseEdgeExistsFilter.create(
                reverseEdge,
                headFilterInputType.filter(value, context, path),
              ),
            ),
        },
      ),
      ...(['eq', 'not', 'gt', 'gte', 'lt', 'lte'] as const).map(
        (operator) =>
          new ReverseEdgeFilterInput<number>(reverseEdge, `count_${operator}`, {
            name:
              operator === 'eq'
                ? reverseEdge.countFieldName
                : `${reverseEdge.countFieldName}_${operator}`,
            type: new utils.NonNullableInputType(
              scalars.typesByName.UnsignedInt,
            ),
            filter: (value, _context, _path) =>
              operator === 'not'
                ? NotOperation.create(
                    MultipleReverseEdgeCountFilter.create(
                      reverseEdge,
                      'eq',
                      value,
                    ),
                  )
                : MultipleReverseEdgeCountFilter.create(
                    reverseEdge,
                    operator,
                    value,
                  ),
          }),
      ),
    ];
  }

  public constructor(
    public readonly node: Node,
    override?: Partial<NodeFilterInputTypeOverride>,
  ) {
    super({
      name: override?.name ?? `${node}FilterInput`,
      description: override?.description ?? `The "${node}" nodes' filter`,
    });
  }

  public createBooleanOperationFields(): BooleanOperationFilterInput[] {
    return [
      new BooleanOperationFilterInput<NodeFilterInputValue[]>({
        name: 'AND',
        type: new utils.NonNullableInputType(new utils.ListableInputType(this)),
        filter: (values, context, path) =>
          AndOperation.create(
            utils.aggregateGraphError<NodeFilterInputValue, BooleanFilter[]>(
              values,
              (operands, value, index) => [
                ...operands,
                this.filter(value, context, utils.addPath(path, index)).filter,
              ],
              [],
              { path },
            ),
          ),
      }),
      new BooleanOperationFilterInput<NodeFilterInputValue[]>({
        name: 'OR',
        type: new utils.NonNullableInputType(new utils.ListableInputType(this)),
        filter: (values, context, path) =>
          OrOperation.create(
            utils.aggregateGraphError<NodeFilterInputValue, BooleanFilter[]>(
              values,
              (operands, value, index) => [
                ...operands,
                this.filter(value, context, utils.addPath(path, index)).filter,
              ],
              [],
              { path },
            ),
          ),
      }),
      new BooleanOperationFilterInput<NodeFilterInputValue>({
        name: 'NOT',
        type: this,
        filter: (value, context, path) =>
          NotOperation.create(this.filter(value, context, path).filter),
      }),
    ];
  }

  @Memoize()
  public override get fields(): ReadonlyArray<FieldFilterInputInterface> {
    const constructor = this.constructor as typeof NodeFilterInputType;

    return [
      ...Array.from(this.node.componentSet).flatMap<FieldFilterInputInterface>(
        (component) =>
          component instanceof Leaf
            ? constructor.createLeafFields(component)
            : constructor.createEdgeFields(component),
      ),
      ...Array.from(
        this.node.reverseEdgeSet,
      ).flatMap<FieldFilterInputInterface>((reverseEdge) =>
        reverseEdge instanceof UniqueReverseEdge
          ? constructor.createUniqueReverseEdgeFields(reverseEdge)
          : constructor.createMultipleReverseEdgeFields(reverseEdge),
      ),
      ...this.createBooleanOperationFields(),
    ];
  }

  public filter(
    value: NodeFilterInputValue,
    context?: OperationContext,
    path?: utils.Path,
  ): NodeFilter {
    return new NodeFilter(
      this.node,
      value === undefined
        ? TrueValue
        : value === null
          ? FalseValue
          : AndOperation.create(
              Object.entries(value).map(([filterName, filterValue]) =>
                this.getFieldByName(filterName, path).filter(
                  filterValue,
                  context,
                  utils.addPath(path, filterName),
                ),
              ),
            ),
    );
  }

  public parseAndFilter(
    maybeValue: unknown,
    context?: OperationContext,
    path?: utils.Path,
  ): NodeFilter {
    return this.filter(this.parseValue(maybeValue, path), context, path);
  }
}
