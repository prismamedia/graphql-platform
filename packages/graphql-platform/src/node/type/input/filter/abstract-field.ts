import * as utils from '@prismamedia/graphql-platform-utils';
import type { Except } from 'type-fest';
import type { OperationContext } from '../../../operation/context.js';
import type { BooleanFilter } from '../../../statement/filter/boolean.js';

export type AbstractFieldFilterInputConfig<TValue> = Except<
  utils.InputConfig<TValue>,
  'defaultValue' | 'parser'
> & {
  filter(
    value: Exclude<TValue, undefined>,
    context: OperationContext | undefined,
    path: utils.Path,
  ): BooleanFilter;
};

export abstract class AbstractFieldFilterInput<
  TValue = any,
> extends utils.Input<TValue> {
  public readonly filter: AbstractFieldFilterInputConfig<TValue>['filter'];

  public constructor({
    filter,
    ...config
  }: AbstractFieldFilterInputConfig<TValue>) {
    super(config);

    this.filter = filter;
  }
}
