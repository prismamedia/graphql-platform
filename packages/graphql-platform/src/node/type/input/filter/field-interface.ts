import type * as utils from '@prismamedia/graphql-platform-utils';
import type { OperationContext } from '../../../operation.js';
import type { BooleanFilter } from '../../../statement.js';

export interface FieldFilterInputInterface<TValue = any>
  extends utils.Input<TValue> {
  filter(
    value: Exclude<TValue, undefined>,
    context: OperationContext | undefined,
    path: utils.Path,
  ): BooleanFilter;
}
