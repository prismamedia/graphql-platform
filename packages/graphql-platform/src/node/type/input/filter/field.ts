export * from './abstract-field.js';
export * from './field/boolean-operation.js';
export * from './field/edge.js';
export * from './field/leaf.js';
export * from './field/reverse-edge.js';
