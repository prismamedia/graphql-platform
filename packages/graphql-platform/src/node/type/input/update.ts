import * as utils from '@prismamedia/graphql-platform-utils';
import { Memoize } from '@prismamedia/ts-memoize';
import type { Except } from 'type-fest';
import type { Node, NodeValue } from '../../../node.js';
import type { MutationContext } from '../../operation.js';
import type { NodeUpdateValue } from '../../statement/update.js';
import {
  ComponentUpdateInput,
  FieldUpdateInput,
  LeafUpdateInput,
  ReverseEdgeUpdateInput,
} from './update/field.js';

export * from './update/field.js';

export type NodeUpdateInputValue = utils.Nillable<utils.PlainObject>;

export class NodeUpdateInputType extends utils.ObjectInputType<FieldUpdateInput> {
  public constructor(public readonly node: Node) {
    super({
      name: `${node}UpdateInput`,
      description: `The "${node}" node's ${utils.MutationType.UPDATE}`,
    });
  }

  @Memoize()
  protected get componentFields(): ReadonlyArray<ComponentUpdateInput> {
    return this.node.components.reduce<ComponentUpdateInput[]>(
      (fields, component) =>
        component.isMutable() ? [...fields, component.updateInput] : fields,
      [],
    );
  }

  @Memoize()
  protected get reverseEdgeFields(): ReadonlyArray<ReverseEdgeUpdateInput> {
    return this.node.reverseEdges.reduce<ReverseEdgeUpdateInput[]>(
      (fields, reverseEdge) =>
        reverseEdge.updateInput ? [...fields, reverseEdge.updateInput] : fields,
      [],
    );
  }

  @Memoize()
  protected get virtualFields(): ReadonlyArray<utils.Input> {
    const { config, configPath } = this.node.getMutationConfig(
      utils.MutationType.UPDATE,
    );

    const virtualFieldsConfig = config?.virtualFields;
    const virtualFieldsConfigPath = utils.addPath(configPath, 'virtualFields');

    utils.assertNillablePlainObjectConfig(
      virtualFieldsConfig,
      virtualFieldsConfigPath,
    );

    return virtualFieldsConfig
      ? utils.aggregateConfigError<
          [utils.Name, Except<utils.InputConfig, 'name'>],
          utils.Input[]
        >(
          Object.entries(virtualFieldsConfig),
          (fields, [virtualFieldName, virtualFieldConfig]) => {
            if (this.node.componentsByName.has(virtualFieldName)) {
              throw new utils.UnexpectedConfigError(
                `not to have a component's name`,
                virtualFieldName,
                { path: virtualFieldsConfigPath },
              );
            } else if (this.node.reverseEdgesByName.has(virtualFieldName)) {
              throw new utils.UnexpectedConfigError(
                `not to have a reverse-edge's name`,
                virtualFieldName,
                { path: virtualFieldsConfigPath },
              );
            }

            return [
              ...fields,
              new utils.Input(
                { ...virtualFieldConfig, name: virtualFieldName },
                utils.addPath(virtualFieldsConfigPath, virtualFieldName),
              ),
            ];
          },
          [],
          { path: virtualFieldsConfigPath },
        )
      : [];
  }

  @Memoize()
  public override get fields(): ReadonlyArray<FieldUpdateInput> {
    return [
      ...this.componentFields,
      ...this.reverseEdgeFields,
      ...this.virtualFields,
    ];
  }

  @Memoize()
  public override isPublic(): boolean {
    return (
      this.node.isMutationPublic(utils.MutationType.UPDATE) &&
      super.isPublic() &&
      this.componentFields.some((field) => field.isPublic())
    );
  }

  public async resolveValue(
    data: Readonly<NonNullable<NodeUpdateInputValue>>,
    context: MutationContext,
    path: utils.Path = utils.addPath(undefined, this.name),
  ): Promise<NodeUpdateValue> {
    const resolvedValue: NodeUpdateValue = Object.create(null);

    await Promise.all(
      this.componentFields.map(async (field) => {
        const fieldData = data[field.name];

        resolvedValue[field.name] =
          fieldData == null || field instanceof LeafUpdateInput
            ? fieldData
            : await field.resolveValue(
                fieldData,
                context,
                utils.addPath(path, field.name),
              );
      }),
    );

    return resolvedValue;
  }

  public hasReverseEdgeActions(
    data: Readonly<NonNullable<NodeUpdateInputValue>>,
  ): boolean {
    return this.reverseEdgeFields.some((field) => data[field.name] != null);
  }

  public async applyReverseEdgeActions(
    nodeValue: NodeValue,
    data: Readonly<NonNullable<NodeUpdateInputValue>>,
    context: MutationContext,
    path?: utils.Path,
  ): Promise<void> {
    await Promise.all(
      this.reverseEdgeFields.map(async (field) => {
        const fieldData = data[field.name];

        if (fieldData != null) {
          await field.applyActions(
            nodeValue,
            fieldData,
            context,
            utils.addPath(path, field.name),
          );
        }
      }),
    );
  }
}
