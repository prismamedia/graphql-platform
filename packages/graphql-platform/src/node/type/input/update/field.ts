import type { Input } from '@prismamedia/graphql-platform-utils';
import type { ComponentUpdateInput } from './field/component.js';
import type { ReverseEdgeUpdateInput } from './field/reverse-edge.js';

export * from './field/component.js';
export * from './field/reverse-edge.js';

export type FieldUpdateInput =
  | ComponentUpdateInput
  | ReverseEdgeUpdateInput
  | Input;
