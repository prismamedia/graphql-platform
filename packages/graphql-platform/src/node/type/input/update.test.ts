import * as utils from '@prismamedia/graphql-platform-utils';
import { GraphQLInputObjectType, printType } from 'graphql';
import { GraphQLPlatform } from '../../../index.js';
import {
  MyGP,
  myUserContext,
  nodeNames,
  nodes,
} from '../../../__tests__/config.js';
import { mockConnector } from '../../../__tests__/connector-mock.js';
import { MutationContext } from '../../operation.js';
import { LeafComparisonFilter } from '../../statement/filter.js';
import { NodeUpdate } from '../../statement/update.js';
import { NodeUpdateInputType } from './update.js';

describe('NodeUpdateInputType', () => {
  let gp: MyGP;

  beforeAll(() => {
    gp = new GraphQLPlatform({
      nodes,
      connector: mockConnector({
        find: async ({ node, filter }) => {
          if (
            node.name === 'Category' &&
            filter?.filter.equals(
              new LeafComparisonFilter(
                node.getLeafByName('id'),
                'eq',
                '91a7c846-b030-4ef3-aaaa-747fe7b11519',
              ),
            )
          ) {
            return [{ _id: 4 }];
          } else if (
            node.name === 'User' &&
            filter?.filter.equals(
              new LeafComparisonFilter(
                node.getLeafByName('id'),
                'eq',
                '2059b77a-a735-41fe-b415-5b12944b6ba6',
              ),
            )
          ) {
            return [
              { id: '2059b77a-a735-41fe-b415-5b12944b6ba6', username: 'yvann' },
            ];
          }

          throw new Error('KO');
        },
      }),
    });
  });

  describe('Definition', () => {
    it.each(nodeNames)('%s has an update input type', (nodeName) => {
      const node = gp.getNodeByName(nodeName);

      const updateInputType = node.updateInputType;
      expect(updateInputType).toBeInstanceOf(NodeUpdateInputType);

      if (node.isMutationPublic(utils.MutationType.UPDATE)) {
        expect(updateInputType.getGraphQLInputType()).toBeInstanceOf(
          GraphQLInputObjectType,
        );

        expect(
          printType(updateInputType.getGraphQLInputType()),
        ).toMatchSnapshot(updateInputType.name);
      } else {
        expect(() => updateInputType.getGraphQLInputType()).toThrowError(
          `The "${nodeName}UpdateInput" input type is private`,
        );
      }
    });
  });

  describe('Runtime', () => {
    describe('Fails', () => {
      it.each([
        [
          'Article',
          {
            category: {},
          },
          '"ArticleUpdateInput.category" - Expects one and only one action, got: [Object: null prototype] {}',
        ],
        [
          'Article',
          {
            category: {
              connect: { id: '91a7c846-b030-4ef3-aaaa-747fe7b11519' },
              connectIfExists: { id: '91a7c846-b030-4ef3-aaaa-747fe7b11519' },
            },
          },
          '"ArticleUpdateInput.category" - Expects one and only one action, got: [Object: null prototype] {',
        ],
      ])(
        '%sUpdateInput.parseValue(%p) throws an error',
        (nodeName, value, error) => {
          const node = gp.getNodeByName(nodeName);
          const updateInputType = node.updateInputType;

          expect(() => updateInputType.parseValue(value)).toThrowError(error);
        },
      );
    });

    describe('Works', () => {
      it.each([
        [
          'Article',
          {},
          {
            updatedAt: expect.any(Date),
          },
        ],
        [
          'Article',
          {
            title: 'My new title',
            category: null,
          },
          {
            title: 'My new title',
            category: null,
            updatedAt: expect.any(Date),
          },
        ],
      ])('%sUpdateInput.parseValue(%p)', (nodeName, value, update) => {
        const node = gp.getNodeByName(nodeName);
        const updateInputType = node.updateInputType;

        expect(updateInputType.parseValue(value)).toEqual(update);
      });

      it('creates a valid statement', async () => {
        const Article = gp.getNodeByName('Article');
        const ArticleUpdateInputType = Article.updateInputType;

        const input: utils.PlainObject = {
          title: "My new article's title",
          category: { disconnect: true },
        };

        const parsedInput = ArticleUpdateInputType.parseValue(input);

        expect(parsedInput).toEqual({
          title: "My new article's title",
          category: { disconnect: true },
          updatedAt: expect.any(Date),
        });

        const update = await ArticleUpdateInputType.createStatement(
          parsedInput!,
          new MutationContext(gp, myUserContext),
        );

        expect(update).toBeInstanceOf(NodeUpdate);

        const proxy = update.proxy;
        expect(proxy).toBeInstanceOf(NodeUpdate);
        expect({ ...proxy }).toEqual({
          title: "My new article's title",
          category: null,
          updatedAt: expect.any(Date),
        });
      });
    });
  });
});
