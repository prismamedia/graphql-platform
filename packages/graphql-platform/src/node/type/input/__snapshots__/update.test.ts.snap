// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NodeUpdateInputType Definition Article may have an update input type: ArticleUpdateInput 1`] = `
""""The "Article" node's update"""
input ArticleUpdateInput {
  status: ArticleStatus
  title: NonEmptySanitizedString
  slug: Slug

  """The article's body"""
  body: DraftJS
  category: ArticleUpdateCategoryInput
  updatedBy: ArticleUpdateUpdatedByInput

  """Contains any arbitrary data you want to store alongside the article"""
  metas: JSONObject

  """Is the article highlighted?"""
  highlighted: Boolean

  """Is the article a partnership?"""
  sponsored: Boolean
  views: UnsignedBigInt
  score: UnsignedFloat
  machineTags: JSONArray
  tags: ArticleUpdateTagsInput
  extension: ArticleUpdateExtensionInput

  """It is possible to provide the article's body as raw HTML"""
  htmlBody: String
}"
`;

exports[`NodeUpdateInputType Definition Article may have an update input type: ArticleUpdateWithoutCategoryInput 1`] = `
""""The "Article" node's update"""
input ArticleUpdateWithoutCategoryInput {
  status: ArticleStatus
  title: NonEmptySanitizedString
  slug: Slug

  """The article's body"""
  body: DraftJS
  updatedBy: ArticleUpdateUpdatedByInput

  """Contains any arbitrary data you want to store alongside the article"""
  metas: JSONObject

  """Is the article highlighted?"""
  highlighted: Boolean

  """Is the article a partnership?"""
  sponsored: Boolean
  views: UnsignedBigInt
  score: UnsignedFloat
  machineTags: JSONArray
  tags: ArticleUpdateTagsInput
  extension: ArticleUpdateExtensionInput

  """It is possible to provide the article's body as raw HTML"""
  htmlBody: String
}"
`;

exports[`NodeUpdateInputType Definition Article may have an update input type: ArticleUpdateWithoutCreatedByInput 1`] = `
""""The "Article" node's update"""
input ArticleUpdateWithoutCreatedByInput {
  status: ArticleStatus
  title: NonEmptySanitizedString
  slug: Slug

  """The article's body"""
  body: DraftJS
  category: ArticleUpdateCategoryInput
  updatedBy: ArticleUpdateUpdatedByInput

  """Contains any arbitrary data you want to store alongside the article"""
  metas: JSONObject

  """Is the article highlighted?"""
  highlighted: Boolean

  """Is the article a partnership?"""
  sponsored: Boolean
  views: UnsignedBigInt
  score: UnsignedFloat
  machineTags: JSONArray
  tags: ArticleUpdateTagsInput
  extension: ArticleUpdateExtensionInput

  """It is possible to provide the article's body as raw HTML"""
  htmlBody: String
}"
`;

exports[`NodeUpdateInputType Definition Article may have an update input type: ArticleUpdateWithoutUpdatedByInput 1`] = `
""""The "Article" node's update"""
input ArticleUpdateWithoutUpdatedByInput {
  status: ArticleStatus
  title: NonEmptySanitizedString
  slug: Slug

  """The article's body"""
  body: DraftJS
  category: ArticleUpdateCategoryInput

  """Contains any arbitrary data you want to store alongside the article"""
  metas: JSONObject

  """Is the article highlighted?"""
  highlighted: Boolean

  """Is the article a partnership?"""
  sponsored: Boolean
  views: UnsignedBigInt
  score: UnsignedFloat
  machineTags: JSONArray
  tags: ArticleUpdateTagsInput
  extension: ArticleUpdateExtensionInput

  """It is possible to provide the article's body as raw HTML"""
  htmlBody: String
}"
`;

exports[`NodeUpdateInputType Definition ArticleExtension may have an update input type: ArticleExtensionUpdateInput 1`] = `
""""The "ArticleExtension" node's update"""
input ArticleExtensionUpdateInput {
  """The source of the article"""
  source: NonEmptyTrimmedString
}"
`;

exports[`NodeUpdateInputType Definition ArticleExtension may have an update input type: ArticleExtensionUpdateWithoutArticleInput 1`] = `
""""The "ArticleExtension" node's update"""
input ArticleExtensionUpdateWithoutArticleInput {
  """The source of the article"""
  source: NonEmptyTrimmedString
}"
`;

exports[`NodeUpdateInputType Definition ArticleTag may have an update input type: ArticleTagUpdateInput 1`] = `
""""The "ArticleTag" node's update"""
input ArticleTagUpdateInput {
  order: UnsignedInt
  moderations: ArticleTagUpdateModerationsInput
}"
`;

exports[`NodeUpdateInputType Definition ArticleTag may have an update input type: ArticleTagUpdateWithoutArticleInput 1`] = `
""""The "ArticleTag" node's update"""
input ArticleTagUpdateWithoutArticleInput {
  order: UnsignedInt
  moderations: ArticleTagUpdateModerationsInput
}"
`;

exports[`NodeUpdateInputType Definition ArticleTag may have an update input type: ArticleTagUpdateWithoutTagInput 1`] = `
""""The "ArticleTag" node's update"""
input ArticleTagUpdateWithoutTagInput {
  order: UnsignedInt
  moderations: ArticleTagUpdateModerationsInput
}"
`;

exports[`NodeUpdateInputType Definition ArticleTagModeration may have an update input type: ArticleTagModerationUpdateInput 1`] = `
""""The "ArticleTagModeration" node's update"""
input ArticleTagModerationUpdateInput {
  moderation: NonEmptyTrimmedString
}"
`;

exports[`NodeUpdateInputType Definition ArticleTagModeration may have an update input type: ArticleTagModerationUpdateWithoutArticleTagInput 1`] = `
""""The "ArticleTagModeration" node's update"""
input ArticleTagModerationUpdateWithoutArticleTagInput {
  moderation: NonEmptyTrimmedString
}"
`;

exports[`NodeUpdateInputType Definition ArticleTagModeration may have an update input type: ArticleTagModerationUpdateWithoutModeratorInput 1`] = `
""""The "ArticleTagModeration" node's update"""
input ArticleTagModerationUpdateWithoutModeratorInput {
  moderation: NonEmptyTrimmedString
}"
`;

exports[`NodeUpdateInputType Definition Category may have an update input type: CategoryUpdateInput 1`] = `
""""The "Category" node's update"""
input CategoryUpdateInput {
  parent: CategoryUpdateParentInput
  order: UnsignedInt

  """This category's children"""
  children: CategoryUpdateChildrenInput

  """The articles attached to this category"""
  articles: CategoryUpdateArticlesInput
}"
`;

exports[`NodeUpdateInputType Definition Category may have an update input type: CategoryUpdateWithoutParentInput 1`] = `
""""The "Category" node's update"""
input CategoryUpdateWithoutParentInput {
  order: UnsignedInt

  """This category's children"""
  children: CategoryUpdateChildrenInput

  """The articles attached to this category"""
  articles: CategoryUpdateArticlesInput
}"
`;

exports[`NodeUpdateInputType Definition Tag may have an update input type: TagUpdateInput 1`] = `
""""The "Tag" node's update"""
input TagUpdateInput {
  """A tag can be deprecated"""
  deprecated: Boolean
  articles: TagUpdateArticlesInput
}"
`;

exports[`NodeUpdateInputType Definition UserProfile may have an update input type: UserProfileUpdateInput 1`] = `
""""The "UserProfile" node's update"""
input UserProfileUpdateInput {
  birthday: Date
  facebookId: ID
  googleId: ID
  twitterHandle: ID
}"
`;

exports[`NodeUpdateInputType Definition UserProfile may have an update input type: UserProfileUpdateWithoutUserInput 1`] = `
""""The "UserProfile" node's update"""
input UserProfileUpdateWithoutUserInput {
  birthday: Date
  facebookId: ID
  googleId: ID
  twitterHandle: ID
}"
`;
